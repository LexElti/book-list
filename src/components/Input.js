import React from 'react'
import {connect} from 'react-redux'
import {changeFilters} from '../store/actions'

export const Input = ({name, filters, changeFilters}) => {
  const changeHandler = event => {
    event.persist()
    const field = event.target.name 
    const value = event.target.value
    changeFilters(field, value)
  }

  return (
    <div className="col-lg-3  col-sm-6 mb-4">
      <input 
        type="text" 
        className="form-control" 
        name={name}
        value={filters[name]}
        placeholder={name.charAt(0).toUpperCase() + name.slice(1)}
        onChange={changeHandler}
      />
    </div>
  )
}

const mapStateToProps = state => ({
  filters: state.books.filters
})

const mapDispatchToProps = {
  changeFilters
}

export default connect(mapStateToProps, mapDispatchToProps)(Input)
