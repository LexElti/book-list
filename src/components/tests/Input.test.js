import React from 'react'
import {Input} from '../Input'

describe('<Input />', () => {
  const props = {
    name: 'title',
    filters: {
      title: '', author: '', publisher: '', subject: ''
    },
    changeFilters: () => {},
  }
  
  describe('<Input /> initial', () => {
    const wrapper = shallow(<Input {...props} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render component with props', () => {
      const input = wrapper.find('input')
      expect(input.prop('name')).toEqual('title')
      expect(input.prop('value')).toEqual('')
      expect(input.prop('placeholder')).toEqual('Title')
    })
  })

  describe('<Input /> some change', () => {
    const nextProps = {
      ...props,
      filters: {
        ...props.filters,
        title: 'Some title'
      }
    }

    const wrapper = shallow(<Input {...nextProps} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render component with props', () => {
      const input = wrapper.find('input')
      expect(input.prop('name')).toEqual('title')
      expect(input.prop('value')).toEqual('Some title')
      expect(input.prop('placeholder')).toEqual('Title')
    })
  })

  describe('Input handler', () => {
    const mockFetchBooks = jest.fn()
    
    const nextProps = {
      ...props,
      changeFilters: mockFetchBooks
    }

    const wrapper = shallow(<Input {...nextProps} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('when submiting the form', () => {
      beforeEach(() => {
        wrapper.find('input').simulate('change', {
          persist: () => {},
          target: {
            value: 'Some title'
          }
        })
      })

      it('calls method changeFilters', () => {
        expect(mockFetchBooks).toHaveBeenCalledTimes(1)
      })
    })
  })
})
