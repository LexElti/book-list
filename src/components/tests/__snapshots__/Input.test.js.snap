// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> <Input /> initial renders properly 1`] = `ShallowWrapper {}`;

exports[`<Input /> <Input /> some change renders properly 1`] = `ShallowWrapper {}`;

exports[`<Input /> Input handler renders properly 1`] = `ShallowWrapper {}`;
